bookkeeper-server/src/main/java/org/apache/bookkeeper/client/ForceLedgerOp.java

```diff
@@ -0,0 +1,123 @@
+
+    DistributionSchedule.AckSet ackSet;
+    boolean completed = false;
+    int lastSeenError = BKException.Code.WriteException;
+    ArrayList<BookieSocketAddress> currentEnsemble;
+
+    long currentNonDurableLastAddConfirmed = LedgerHandle.INVALID_ENTRY_ID;
+
+    final LedgerHandle lh;
+
+    ForceLedgerOp(LedgerHandle lh, CompletableFuture<Void> cb) {
+        this.lh = lh;
+        this.cb = cb;
+    }
+
+    void sendForceLedgerRequest(int bookieIndex) {
+        lh.bk.getBookieClient().forceLedger(currentEnsemble.get(bookieIndex), lh.ledgerId, this, bookieIndex);
+    }
+
+    @Override
+    public void safeRun() {
+        initiate();
+    }
+
+    void initiate() {
+
+        // capture currentNonDurableLastAddConfirmed
+        // remember that we are inside OrderedExecutor, this induces a strict ordering
+        // on the sequence of events
+        this.currentNonDurableLastAddConfirmed = lh.pendingAddsSequenceHead;
+        LOG.debug("force {} clientNonDurableLac {}", lh.ledgerId, currentNonDurableLastAddConfirmed);
+
+        // we need to send the request to every bookie in the ensamble
+        this.currentEnsemble = lh.metadata.currentEnsemble;
+        this.ackSet = lh.distributionSchedule.getAckSetForForceLedger();
+        this.receivedResponseSet = new BitSet(
+                lh.getLedgerMetadata().getEnsembleSize());
+        this.receivedResponseSet.set(0,
+                lh.getLedgerMetadata().getEnsembleSize());
+
+        DistributionSchedule.WriteSet writeSet = lh.distributionSchedule.getWriteSetForForceLedger();
+        try {
+            for (int i = 0; i < writeSet.size(); i++) {
+                sendForceLedgerRequest(writeSet.get(i));
+            }
+        } finally {
+            writeSet.recycle();
+        }
+    }
+
+    @Override
+    public void forceLedgerComplete(int rc, long ledgerId, BookieSocketAddress addr, Object ctx) {
+        int bookieIndex = (Integer) ctx;
+
+        if (completed) {
+            return;
+        }
+
+        if (BKException.Code.OK != rc) {
+            lastSeenError = rc;
+        }
+
+        // We got response.
+        receivedResponseSet.clear(bookieIndex);
+
+        if (rc == BKException.Code.OK) {
+            if (ackSet.completeBookieAndCheck(bookieIndex) && !completed) {
+                completed = true;
+                // we are able to say that every bookie sync'd its own journal
+                // for every ackknowledged entry before issuing the force() call
+                LOG.debug("After force on ledger {} updating LastAddConfirmed to {} ",
+                        ledgerId, currentNonDurableLastAddConfirmed);
+                lh.lastAddConfirmed = currentNonDurableLastAddConfirmed;
+                FutureUtils.complete(cb, null);
+                return;
+            }
+        } else {
+            LOG.warn("ForceLedger did not succeed: Ledger {} on {}", new Object[]{ledgerId, addr});
+        }
+
+        if (receivedResponseSet.isEmpty()) {
```

Comment:
Ok; I am little confused with the logic here. We need to have the following logic.
1. Success case: we set `completed` flag when all bookies send  `Code.OK` and `receivedResponseSet.isEmpty`.
2. Failure case:  Set `completed` flag on the first error and ignore rest of the responses. 
Why do you need to wait until all responses in the failure case?

Reply:
good idea. reworked